java/src/main/java/com/mlt/converter/encodings/fsst/Fsst.java

```diff
@@ -0,0 +1,46 @@
+package com.mlt.converter.encodings.fsst;
+
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.nio.ByteOrder;
+
+public interface Fsst {
+  SymbolTable encode(byte[] data);
+
+  default byte[] decode(SymbolTable encoded) throws IOException {
+    return decode(encoded.symbols(), encoded.symbolLengths(), encoded.compressedData());
+  }
+
+  default byte[] decode(byte[] symbols, int[] symbolLengths, byte[] compressedData)
+      throws IOException {
+    ByteArrayOutputStream decodedData = new ByteArrayOutputStream();
+    ByteBuffer symbolsBuffer = ByteBuffer.wrap(symbols).order(ByteOrder.BIG_ENDIAN);
+
+    int[] symbolOffsets = new int[symbolLengths.length];
+    for (int i = 1; i < symbolLengths.length; i++) {
+      symbolOffsets[i] = symbolOffsets[i - 1] + symbolLengths[i - 1];
+    }
+
+    for (int i = 0; i < compressedData.length; i++) {
+      // In Java a byte[] is signed [-128 to 127], whereas in C++ it is unsigned [0 to 255]
+      // So we do a bit shifting operation to convert the values into unsigned values for easier
+      // handling
+      int symbolIndex = compressedData[i] & 0xFF;
+      // 255 is our escape byte -> take the next symbol as it is
+      if (symbolIndex == 255) {
+        decodedData.write(compressedData[++i] & 0xFF);
+      } else if (symbolIndex < symbolLengths.length) {
+        int symbolLength = symbolLengths[symbolIndex];
+        int symbolOffset = symbolOffsets[symbolIndex];
+        byte[] symbolBytes = new byte[symbolLength];
+        for (int j = 0; j < symbolLength; j++) {
```

Comment:
I think there is no need for using a ByteBuffer class in that case for the symbolBuffer, so we could get rid of that extra loop by directly writing to the ByteArrayOutputStream with an specified offset or am I missing something?

Reply:
In general in MLT the size of the decoded buffer is also part of the tile, so that we can preallocate the buffer. Therefore, we could get rid of the dynamic allocation overhead based on the  ByteArrayOutputStream

Reply:
Good catch, I copied decoder over without any changes but it's a lot cleaner now with that fix.  I added a benchmark for that as well, looks like it's about 4x faster.

I added a new `decode` method that takes an expected length, but I couldn't figure out how to get that expected length in the callers so I just left the old one too and marked it as deprecated so we can update to use the new one in a followup PR?

Reply:
Yes, we can change it in the next PR